### ISSUE TITLE
Cache-Control: no-cache breaks the middleware

### ISSUE THREAD

**Issue by tinchou** [aspnet/ResponseCaching#81]
**TL;DR**: this breaks with the `Cache-Control: no-cache` header since the `IResponseCachingFeature` feature is not set up and we hit [this exception](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/Internal/ResponseCacheFilter.cs#L132).

-------

Hi,

I'm just starting with your middleware, and the basic functionality is great. I just specify a `[ResponseCache]` and it gets cached automatically, which is awesome :).

**But** I'm getting an error 500 each time I do "Ctrl+F5" or I disable caching in the Chrome dev tools. The error is the one in the title:

```
An unhandled exception occurred while processing the request. InvalidOperationException: 'VaryByQueryKeys' requires the response cache middleware.

Microsoft.AspNetCore.Mvc.Internal.ResponseCacheFilter.OnActionExecuting(ActionExecutingContext context)
```

The error is the same as if I had not called `app.UseResponseCaching();` or called it in the wrong order. The exact line is the following, which shows that it's checking for a `Feature` of type `IResponseCachingFeature`: https://github.com/aspnet/Mvc/blob/d8c6c4ab34e1368c1b071a01fcdcb9e8cc12e110/src/Microsoft.AspNetCore.Mvc.Core/Internal/ResponseCacheFilter.cs#L132.

I investigated the source code and came to the conclusion that the error is produced because:

1. Browsers send `Cache-Control: no-cache` header
1. The request is marked as not cacheable [here](https://github.com/aspnet/ResponseCaching/blob/f8831928de92f7900f26b74f64eb4083227f257a/src/Microsoft.AspNetCore.ResponseCaching/Internal/ResponseCachingPolicyProvider.cs#L35)
1. Since it's not cacheable we skip the [call to `ShimResponseStream`](https://github.com/aspnet/ResponseCaching/blob/f8831928de92f7900f26b74f64eb4083227f257a/src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs#L87)
1. And we [don't set up the Feature](https://github.com/aspnet/ResponseCaching/blob/2eb66dc684faeec5085cb957fbe5c4e5a11ec4ec/src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs#L340) for the next middlewares.
1. So we end up with the mentioned exception in [this line](https://github.com/aspnet/Mvc/blob/d8c6c4ab34e1368c1b071a01fcdcb9e8cc12e110/src/Microsoft.AspNetCore.Mvc.Core/Internal/ResponseCacheFilter.cs#L132)

If you could give me any workaround until a fix is published I'd really appreciate it!

Thanks,
Martín.

**Comment by Tratcher** [comment]
@tinchou Thanks for the detailed investigation. We'll get this dealt with ASAP. One possible workaround would be to always set an `IResponseCachingFeature`. Try putting this between the caching middleware and MVC:
```
app.Use((context, next) =>
{
  if (context.Features.Get<IResponseCachingFeature>() == null)
  {
    context.Features.Set<IResponseCachingFeature>(new FakeResponseCachingFeature());
  }
  return next();
});
```

@muratg @DamianEdwards This is a patch candidate bug.

**Comment by tinchou** [comment]
Can confirm the workaround works for me.

I implemented `FakeResponseCachingFeature` as:

```cs
    public class FakeResponseCachingFeature : IResponseCachingFeature
    {
        public string[] VaryByQueryKeys
        {
            get => null;
            set { }
        }
}
```

**Comment by JunTaoLuo** [comment]
@pranavkm we'll need to build a 1.1.1 branch for this.

**Comment by JunTaoLuo** [comment]
@Tratcher and I think the middleware needs to be fixed. The feature should be added regardless of whether the request is cacheable since it is used when generating the response. `Cache-Control` headers are unidirectional and response generated by a request with `no-cache` can be used to update the existing entry in the cache.

**Comment by Tratcher** [comment]
In other words, the request header `Cache-Control: no-cache` says do not retrieve the resource from cache, it does not prohibit caching the new response.

**Comment by tinchou** [comment]
Should clients be telling the server when not to use the cache?

Before using this middleware my implementation was based on IMemoryCache,
and I never thought about checking the headers because it was an
implementation detail of the server.

I've looked around but haven't found specific information.

On Wed, Nov 30, 2016 at 3:09 PM Chris R <notifications@github.com> wrote:

In other words, the request header Cache-Control: no-cache says do not
retrieve the resource from cache, it does not prohibit caching the new
response.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/aspnet/ResponseCaching/issues/81#issuecomment-263948597>,
or mute the thread

**Comment by DamianEdwards** [comment]
@tinchou when it comes to HTTP-based response caching, yes, the client still maintains ability to send cache control headers, thus affecting upstream caching behavior. As a result, you shouldn't rely on whole response caching to control load on your application. Using `IMemoryCache` or the `<cache>` Tag Helper are ways you can cache parts of the your application logic and fragments of HTML responses separately from systems that use HTTP cache control headers.

The old docs for ASP.NET Caching (from the System.Web days) covers the differences quite nicely in [this introduction](https://msdn.microsoft.com/en-us/library/xsbfdd8c.aspx).

All that said, it might still be nice to have the response caching middleware support a mode whereby it ignores all client-based HTTP cache control headers, and instead acts as a completely server-controlled cache.

**Comment by tinchou** [comment]
@DamianEdwards yes, I think that would be my preferred scenario. Since I'm already encoding caching information in my `ResponseCache` attribute, it'd be really nice if I could have server-side caching automatically set up too.

But if the client can void my cache as they please, then that defeats the whole performance story.

**Comment by DamianEdwards** [comment]
Because that's normally how HTTP-based caching works. There's a client, server, and any amount of intermediaries in between, and they all speak HTTP and participate in caching, using the same cache control headers.

We should investigate the caching behavior of the popular edge caching appliances & services, along with other frameworks, to see how they differentiate these concerns.

**Comment by GuardRex** [comment]
The Azure CDN doesn't respect an invalidation attempt using the `Cache-Control` header.

Initial request for a stale image serves from the origin ...

![capture1](https://cloud.githubusercontent.com/assets/1622880/20775737/0ce28d6c-b722-11e6-89ea-7a7abf3054c9.PNG)

```
HTTP/1.1 200 OK
Cache-Control: public,max-age=2592000
Content-Type: image/png
Date: Wed, 30 Nov 2016 23:15:08 GMT
Etag: 0x8D39DF4E61FD969
Last-Modified: Sun, 26 Jun 2016 19:05:54 GMT
Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
x-ms-blob-type: BlockBlob
x-ms-lease-status: unlocked
x-ms-meta-CbModifiedTime: Fri, 24 Oct 2014 23:37:07 GMT
x-ms-request-id: fdf3f5e1-0001-0028-6f5f-4b9658000000
x-ms-version: 2009-09-19
Content-Length: 12787
```
Second request with `Cache-Control: no-cache` is not respected by the CDN and it serves from the edge node. Note `X-Cache: HIT` and `Server` ...

![capture2](https://cloud.githubusercontent.com/assets/1622880/20775742/128ca52c-b722-11e6-90bb-a7ae27337ccb.PNG)

```
HTTP/1.1 200 OK
Accept-Ranges: bytes
Cache-Control: public,max-age=2592000
Content-Type: image/png
Date: Wed, 30 Nov 2016 23:15:13 GMT
Etag: 0x8D39DF4E61FD969
Last-Modified: Sun, 26 Jun 2016 19:05:54 GMT
Server: ECAcc (dfw/56E0)
X-Cache: HIT
x-ms-blob-type: BlockBlob
x-ms-lease-status: unlocked
x-ms-meta-CbModifiedTime: Fri, 24 Oct 2014 23:37:07 GMT
x-ms-request-id: fdf3f5e1-0001-0028-6f5f-4b9658000000
x-ms-version: 2009-09-19
Content-Length: 12787
```

**Comment by ctolkien** [comment]
>The Azure CDN doesn't respect an invalidation attempt using the Cache-Control header.

Neither does CloudFlare.

![image](https://cloud.githubusercontent.com/assets/515955/21068979/e7fa4896-bec9-11e6-959f-e339257c998d.png)

![image](https://cloud.githubusercontent.com/assets/515955/21068993/09e3a6dc-beca-11e6-8c3f-bc267a58d564.png)

So I think this behaviour is typical of edge caches.

**Comment by GuardRex** [comment]
> it might still be nice to have the response caching middleware support a mode whereby it ignores all client-based HTTP cache control headers, and instead acts as a completely server-controlled cache.

:+1: **I like this idea.** If offered, I'll definitely use it.

**Comment by JunTaoLuo** [comment]
Fixed in #88. Will need to back port to 1.1.1 in a separate issue.
